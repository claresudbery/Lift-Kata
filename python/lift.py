from enum import Enum

from attr import dataclass


class Direction(Enum):
    UP = 0
    DOWN = 1


@dataclass
class Call:
    floor: int
    direction: Direction


class Lift:

    def __init__(self, id, floor, doors_open=False, requested_floors=None):
        self.id = id
        self.floor = floor
        self.doors_open = doors_open
        self.requested_floors = list(requested_floors) if requested_floors else []

    def move_to(self, new_floor):
        self.floor = new_floor


class LiftSystem:
    def __init__(self, floors=None, lifts=None, calls=None):
        self.floors = list(floors) if floors else []
        self.lifts = list(lifts) if lifts else []
        self.calls = list(calls) if calls else []

    def calls_for(self, floor):
        return [c for c in self.calls if c.floor == floor]

    def tick(self):
        lift = self.lifts[0]
        if lift.requested_floors:
            request = lift.requested_floors[0]
            if lift.floor != request:
                lift.move_to(request)
            else:
                lift.doors_open = True
                lift.requested_floors = []
        else:
            lift.doors_open = False

        if len(self.calls) > 0:
            if lift.floor != self.calls[0].floor:
                lift.move_to(self.calls[0].floor)

    def tick_fulfil_requests_and_calls(self):
        lift = self.lifts[0]
        if lift.requested_floors:
            request = lift.requested_floors[0]
            if lift.floor != request:
                lift.move_to(request)
            else:
                lift.doors_open = True
                lift.requested_floors = []
        else:
            lift.doors_open = False

        if len(self.calls) > 0:
            if lift.floor != self.calls[0].floor:
                lift.move_to(self.calls[0].floor)

    def tick_open_doors_and_clear_request_and_close_doors(self):
        lift = self.lifts[0]
        if lift.requested_floors:
            request = lift.requested_floors[0]
            if lift.floor != request:
                lift.move_to(request)
            else:
                lift.doors_open = True
                lift.requested_floors = []
        else:
            lift.doors_open = False

    def tick_open_doors_and_clear_request(self):
        lift = self.lifts[0]
        if lift.requested_floors:
            request = lift.requested_floors[0]
            if lift.floor != request:
                lift.move_to(request)
            else:
                lift.doors_open = True
                lift.requested_floors = []

    def tick_move_to_floor(self):
        lift = self.lifts[0]
        if lift.requested_floors:
            request = lift.requested_floors[0]
            lift.move_to(request)

    def tick_unimplemented(self):
        # TODO: implement this method
        pass