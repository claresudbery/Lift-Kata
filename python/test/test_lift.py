from approvaltests import verify

from lift import Lift, LiftSystem, Call, Direction
from lift_printers import print_lifts


def test_initialises_with_two_floors_and_single_lift():
    # Arrange
    lift_a = Lift("A", 0)

    # Act
    lifts = LiftSystem(floors=[0, 1], lifts=[lift_a])

    # Assert
    verify(print_lifts(lifts))


def test_idle_lift_with_request_after_tick_moves_to_requested_floor():
    # Arrange
    lift_a = Lift("A", 0, requested_floors=[1])
    lift_system = LiftSystem(floors=[0, 1], lifts=[lift_a])
    lift_system_output = print_lifts(lift_system)

    # Act
    lift_system_output += tick_and_return_output(lift_system)

    # Assert
    verify(lift_system_output)


def test_idle_lift_that_moved_to_floor_after_tick_opens_doors_and_clears_request():
    # Arrange
    lift_a = Lift("A", 1, requested_floors=[1])
    lift_system = LiftSystem(floors=[0, 1], lifts=[lift_a])
    lift_system_output = print_lifts(lift_system)

    # Act
    lift_system_output += tick_and_return_output(lift_system)

    # Assert
    verify(lift_system_output)


def test_lift_that_finished_request_if_no_new_request_closes_doors():
    # Arrange
    lift_a = Lift("A", 1, doors_open=True)
    lift_system = LiftSystem(floors=[0, 1], lifts=[lift_a])
    lift_system_output = print_lifts(lift_system)

    # Act
    lift_system_output += tick_and_return_output(lift_system)

    # Assert
    verify(lift_system_output)


def test_empty_lift_moves_towards_a_waiting_person():
    # Arrange
    lift_a = Lift("A", 1)
    calls = [Call(0, Direction.UP)]
    lift_system = LiftSystem(floors=[0, 1], lifts=[lift_a], calls=calls)
    lift_system_output = print_lifts(lift_system)

    # Act
    lift_system_output += tick_and_return_output(lift_system)

    # Assert
    verify(lift_system_output)

def tick_and_return_output(lift_system):
    lift_system.tick()
    return "...\n" + print_lifts(lift_system)
