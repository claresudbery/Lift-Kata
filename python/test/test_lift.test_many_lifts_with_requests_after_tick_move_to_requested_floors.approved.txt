3                *  3
2           *       2
1      *            1
0     [A]  [B]  [C] 0
...
3              [*C] 3
2         [*B]      2
1    [*A]           1
0                   0
